prow/pod-utils/gcs/upload.go

```diff
@@ -171,12 +197,51 @@ func (w gcsObjectWriter) ApplyAttributes(attrs *storage.ObjectAttrs) {
 	w.Writer.ObjectAttrs = *attrs
 }
 
+type openerObjectWriter struct {
+	pkgio.Opener
+	Bucket      string
+	Dest        string
+	writeCloser pkgio.WriteCloser
+}
+
+func (w *openerObjectWriter) Write(p []byte) (n int, err error) {
+	if w.writeCloser == nil {
+		w.writeCloser, err = w.Opener.Writer(context.Background(), fmt.Sprintf("%s/%s", w.Bucket, w.Dest))
+		if err != nil {
+			return 0, err
+		}
+	}
+	return w.writeCloser.Write(p)
+}
+
+func (w *openerObjectWriter) Close() error {
+	if w.writeCloser != nil {
+		err := w.writeCloser.Close()
+		w.writeCloser = nil
+		return err
+	}
+	return nil
+}
+
+func (w *openerObjectWriter) SetSize(_ int64) {
+	// ChunkSize is in gocloud only implemented by GCS
+	// but we're not using the openerObjectWriter for GCS
+	// so we don't have to set this
+	return
+}
+
+func (w *openerObjectWriter) ApplyAttributes(_ *storage.ObjectAttrs) {
+	// Cannot be set via the current abstraction of the opener package
+	// because blob.WriterOptions is not exposed
+	return
+}
+
 type localFileWriter struct {
 	filePath string
 	file     *os.File
 }
 
-func (*localFileWriter) SetSize(size int64) {}
+func (*localFileWriter) SetSize(_ int64) {}
```

Comment:
Probably more of a follow-up but this can be removed and replaces with the `openerObjectWriter` I guess?

Reply:
I'm not entirely sure. Currently we use only `os.Create` in the opener and `MkdirAll, OpenFile, Write` in the `localFileWriter`.  I'm not sure if `os.Create` also creates directories if necessary. But this might be something we could add to the implementation in opener

Reply:
If we want, I can extend the file writing of the opener like this and use the opener: 
:+1: :-1: ?

Reply:
Replacing localFileWriter with openerWriter sgtm

We may want to make auto-creation of subdirs an option on either the opener struct or Writer() call

Reply:
I replaced the localFileWriter with the openerObjectWriter. I would suggest always executing MkdirAll for now as sane default. When we / if we have cases where we don't want to create the parent dir I would add a option for it. There are slight differences in the file permissions:
* opener: `return OpenFile(name, O_RDWR|O_CREATE|O_TRUNC, 0666)`
* localFileWriter: `os.OpenFile(w.filePath, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0755)`

Currently I'm just using the one from opener. Is that okay, should I use something else, or should I add an option for it (not sure if it's important in our current use cases as localFileWriter is just a local debug mode and I don't think we have to write executable files with it)?